src/acpi.rs

```diff
@@ -419,3 +424,357 @@ pub fn process_msct() -> (
+                        let entry = iterator as *const ACPI_NFIT_FLUSH_ADDRESS;
+                        parse_nfit_flush_hint_structure(entry);
+                    }
+                    Enum_AcpiNfitType::ACPI_NFIT_TYPE_RESERVED => {
+                        let entry = iterator as *const ACPI_NFIT_PLATFORM_CAPABILITIES;
+                        parse_nfit_platform_capabilities_structure(entry);
+                    }
+                    _ => unreachable!(),
+                }
+                iterator = iterator.add((*header).Length as usize);
+            }
+        } else {
+            debug!("ACPI NFIT Table not found.");
+        }
+        pmem_descriptors
+    }
+}
+
+bitflags! {
+    pub struct MappingAttributes: u64 {
+        const EFI_MEMORY_UC = 0x00000001;
+    const EFI_MEMORY_WC = 0x00000002;
+    const EFI_MEMORY_WT = 0x00000004;
+    const EFI_MEMORY_WB = 0x00000008;
+    const EFI_MEMORY_UCE = 0x00000010;
+    const EFI_MEMORY_WP = 0x00001000;
+    const EFI_MEMORY_RP = 0x00002000;
+    const EFI_MEMORY_XP = 0x00004000;
+    const EFI_MEMORY_NV = 0x00008000;
+    const EFI_MEMORY_MORE_RELIABLE = 0x00010000;
+    const EFI_MEMORY_RO = 0x00020000;
+    const EFI_MEMORY_SP = 0x00040000;
+    }
+}
+
+/// Convert u64 to MappingAttributes.
+impl From<u64> for MappingAttributes {
+    fn from(mode: u64) -> MappingAttributes {
+        MappingAttributes::from_bits_truncate(mode)
+    }
+}
+
+/// Convert MappingAttributes to u64.
+impl From<MappingAttributes> for u64 {
+    fn from(mode: MappingAttributes) -> u64 {
+        mode.bits()
+    }
+}
+
+struct Guid {
+    data1: u32,
+    data2: u16,
+    data3: u16,
+    index8: u8,
+    index9: u8,
+    index10: u8,
+    index11: u8,
+    index12: u8,
+    index13: u8,
+    index14: u8,
+    index15: u8,
+}
+
+impl fmt::Debug for Guid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{{ {:#X}, {:#X}, {:#X}, {:#X}, {:#X}, {:#X}, {:#X}, {:#X}, {:#X}, {:#X}, {:#X} }}",
+            self.data1,
+            self.data2,
+            self.data3,
+            self.index8,
+            self.index9,
+            self.index10,
+            self.index11,
+            self.index12,
+            self.index13,
+            self.index14,
+            self.index15
+        )
+    }
+}
+
+fn from_slice_u8_to_Guid(slice: &[u8]) -> Guid {
+    let p: *const [u8; core::mem::size_of::<Guid>()] =
+        slice.as_ptr() as *const [u8; core::mem::size_of::<Guid>()];
+    unsafe { core::mem::transmute(*p) }
+}
+
+// To parse DeviceHandle in ACPI_NFIT_MEMORY_MAP and ACPI_NFIT_FLUSH_ADDRESS subtables
+const ACPI_NFIT_DEVICE_HANDLE_4BIT_MASK: u32 = 0xf;
+const ACPI_NFIT_DEVICE_HANDLE_12BIT_MASK: u32 = 0xfff;
+
+// NFIT subtable type = 0x0
+fn parse_nfit_spa_range_structure(entry: *const ACPI_NFIT_SYSTEM_ADDRESS) -> MemoryDescriptor {
+    let mut mem_desc = MemoryDescriptor::default();
```

Comment:
Rather than having mut mem_desc, can we create and initialize directly below? e.g. MemoryDescriptor { bla: ... }

Reply:
Previously one field was private in MemoryDescriptor that stopped me from initializing the way you suggested, however, now that a few uefi types are crate local, I am able to do it.

Done.